flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/read/TimestampedHiveInputSplit.java

```diff
@@ -0,0 +1,232 @@
+
+import org.apache.flink.api.common.typeinfo.BasicTypeInfo;
+import org.apache.flink.api.common.typeinfo.TypeInfo;
+import org.apache.flink.api.common.typeinfo.TypeInfoFactory;
+import org.apache.flink.api.common.typeinfo.TypeInformation;
+import org.apache.flink.api.common.typeutils.SimpleTypeSerializerSnapshot;
+import org.apache.flink.api.common.typeutils.TypeSerializerSnapshot;
+import org.apache.flink.api.common.typeutils.base.TypeSerializerSingleton;
+import org.apache.flink.api.java.typeutils.runtime.DataInputViewStream;
+import org.apache.flink.api.java.typeutils.runtime.DataOutputViewStream;
+import org.apache.flink.core.memory.DataInputView;
+import org.apache.flink.core.memory.DataOutputView;
+import org.apache.flink.streaming.api.functions.source.TimestampedInputSplit;
+import org.apache.flink.util.FlinkRuntimeException;
+import org.apache.flink.util.InstantiationUtil;
+
+import java.io.IOException;
+import java.io.Serializable;
+import java.lang.reflect.Type;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * A {@link HiveTableInputSplit} with {@link TimestampedInputSplit}.
+ * Kryo serializer can not deal with hadoop split, need specific type information factory.
+ */
+@TypeInfo(TimestampedHiveInputSplit.SplitTypeInfoFactory.class)
+public class TimestampedHiveInputSplit extends HiveTableInputSplit implements TimestampedInputSplit {
+
+	private static final long serialVersionUID = 1L;
+
+	/** The modification time of the file this split belongs to. */
+	private final long modificationTime;
+
+	/**
+	 * The state of the split. This information is used when
+	 * restoring from a checkpoint and allows to resume reading the
+	 * underlying file from the point we left off.
+	 * */
+	private Serializable splitState;
+
+	public TimestampedHiveInputSplit(
+			long modificationTime,
+			HiveTableInputSplit split) {
+		super(
+				split.getSplitNumber(),
+				split.getHadoopInputSplit(),
+				split.getJobConf(),
+				split.getHiveTablePartition());
+		this.modificationTime = modificationTime;
+	}
+
+	@Override
+	public void setSplitState(Serializable state) {
+		this.splitState = state;
+	}
+
+	@Override
+	public Serializable getSplitState() {
+		return this.splitState;
+	}
+
+	@Override
+	public long getModificationTime() {
+		return modificationTime;
+	}
+
+	@Override
+	public int compareTo(TimestampedInputSplit o) {
+		TimestampedHiveInputSplit split = (TimestampedHiveInputSplit) o;
+		int modTimeComp = Long.compare(this.modificationTime, split.modificationTime);
+		if (modTimeComp != 0L) {
+			return modTimeComp;
+		}
+
+		int pathComp = this.hiveTablePartition.getStorageDescriptor().compareTo(
+				split.hiveTablePartition.getStorageDescriptor());
+
+		return pathComp != 0 ? pathComp :
+				this.getSplitNumber() - o.getSplitNumber();
+	}
+
+	@Override
+	public boolean equals(Object o) {
+		if (this == o) {
+			return true;
+		}
+		if (o == null || getClass() != o.getClass()) {
+			return false;
+		}
+		if (!super.equals(o)) {
```

Comment:
Yes, we can create the case like you said, but in read world, it should not happen.

Reply:
I think the best practice is to avoid such implementations. And if for some reason it has to be done this way, we should clarify it in the java doc, according to the suggestions of `Comparable`:
```
     * <p>It is strongly recommended, but <i>not</i> strictly required that
     * <tt>(x.compareTo(y)==0) == (x.equals(y))</tt>.  Generally speaking, any
     * class that implements the <tt>Comparable</tt> interface and violates
     * this condition should clearly indicate this fact.  The recommended
     * language is "Note: this class has a natural ordering that is
     * inconsistent with equals."
```

Reply:
`TimestampedHiveInputSplit` is only used in `ContinuousFileReaderOperator`. I will add comments.